### ISSUE TITLE
Directional and non-directional service patterns

### ISSUE THREAD

**Issue by scmcca** [MobilityData/gtfs-flex#46]
GTFS-Flex v2 does not currently describe disallowed travel between or within locations. Can [GTFS-TravelRestrictions](https://github.com/google/transit/pull/205) be used in parallel to describe such restrictions in flexible service?

Has anyone ran into this problem or has a solution they currently implement? 

Edit: Renamed issue from "Travel restrictions in GTFS Flex" to "Directional and non-directional service patterns"

**Comment by westontrillium** [comment]
Maybe I'm not thinking critically enough, but can't disallowed pickup/drop-off areas be represented simply by not being included in the location polygon?

If you have a use case, maybe I can model out a solution here.

**Comment by scmcca** [comment]
I was thinking of directionality between zones. Consider ZoneA and ZoneB. Both are on-demand locations that allow travel anywhere inside of them, but the rider may only travel from ZoneA to ZoneB, and not the other way around. Could the directionality be represented using `stop_sequence` for the same `trip_id` (ZoneA has `stop_sequence=1` and ZoneB has `stop_sequence=2`)?

A more general question for processing: how would a consumer know which `trip` to use in the dataset if locations are reused for different trips? Would they have to filter by Origin-Destination pair first, or some other method? What does the processing look like?

**Comment by westontrillium** [comment]
Here's how I would model your example:

> Both are on-demand locations that allow travel anywhere inside of them

Assuming they are part of the same service, each location would get its **own** trip under one route (because there are specific parameters for in-between travel, we can't group them together in the same location_group):

### `stop_times.txt`
trip_id | stop_id | stop_sequence | pickup_type | drop_off_type | start_pickup_dropoff_window | end_pickup_dropoff_window
------ | ------ | ------ | ------ | ------ | ------ | ------
ContiguousZoneA | ZoneA | 1 | 2 | 2 | 08:00:00 | 17:30:00
ContiguousZoneB | ZoneB | 1 | 2 | 2 | 08:00:00 | 17:30:00

Then...
> the rider may only travel from ZoneA to ZoneB, and not the other way around.

A third trip would be created for this unique allowed situation:

trip_id | stop_id | stop_sequence | pickup_type | drop_off_type | start_pickup_dropoff_window | end_pickup_dropoff_window
------ | ------ | ------ | ------ | ------ | ------ | ------
ZoneAtoB | ZoneA | 1 | 2 | 1 | 08:00:00 | 17:30:00
ZoneAtoB | ZoneB | 2 | 1 | 2 | 08:00:00 | 17:30:00

Notice that the pickup/drop-off rules force only one possible origin/destination scenario for this trip (picking up in ZoneA and dropping off in ZoneB), effectively describing the "directionality."

With Flex `trip_ids`, the origin can't be in one trip and the destination be in the other (this isn't unique to Flex, it's just that a fixed route service could use a common `block_id` to have the o/d pair straddle two trips). Thus, the ContiguousZoneA/ContiguousZoneB trips can coexist with the ZoneAtoB trip without duplicative o/ds since they are different trips. Specifically, you won't get a duplicative pickup or drop-off value since an o/d pair contained in either ZoneA or ZoneB can only yield the ContiguousZoneA or ContiguousZoneB trip, an o/d pair of ZoneA to ZoneB can only yield the ZoneAtoB trip, and anything outside these scenarios would be invalid.

> A more general question for processing: how would a consumer know which trip to use in the dataset if locations are reused for different trips? Would they have to filter by Origin-Destination pair first, or some other method? What does the processing look like?

I assume a consumer would whittle down the trip options as it checks the search parameters against the dataset. I don't think there's a particular order that trip parameters need to be filtered in, because the trip has to match **all** of the search parameters (all the parameters are "and" relationships. If any one component is invalid, the trip isn't doable), but here's one possible way.

To use the scenario going from ZoneA to ZoneB:
1. Any locations that the lat/lon of the origin point falls within would be flagged–in this case ZoneA. The list of possible trip_ids is whittled down to trips that include a stop_id of ZoneA.
2. The same logic is used with the destination point, which in our example falls within ZoneB. The list further filters for any trips_ids that also include a stop_id of ZoneB.

At this point, your list only includes trip_ids with **both** stop_id ZoneA and ZoneB.

3. Filter for any trips that include a ZoneA stop_time with any `pickup_type` other than `1` (since the pickup point is in ZoneA).
4. Filter for any trips that include a ZoneB stop_time with any `drop_off_type` other than `1` (since the drop-off point is in ZoneB).
5. Filter for any trips whose ZoneA stop_time's `start/end_pickup_dropoff_window` encompasses the departure time (let's say it's 11:05:00).
6. Generate a travel time estimate from the origin to destination based on the departure time of 11:05:00. Let's say the estimated travel time is 30 minutes to get to the destination. That would mean a drop-off time of 11:35:00. Trips are filtered for any trips whose ZoneB stop_time's `start/end_pickup_dropoff_window` encompasses the estimated drop-off time.

At this point, the consumer could return a itinerary to the rider using what trip_ids are left in the filtered list, since all search parameters have been accounted for.

**Comment by scmcca** [comment]
I hadn't considered that you would need to parse out `trip_id=ZoneAtoB` into separate trips for just `ZoneA` and just `ZoneB`. I would have thought that the rider could travel only within `ZoneA` on the `ZoneAtoB` trip. This would be analogous to if you were travelling between a subset of stops on a fixed transit service. The difference here is that GTFS-Flex allows single `stop_time` trips, which redefines the notion of origin-destination for processing `trip` selection.

A huge benefit to the method you presented above disambiguates any confusion if there are different attributes (i.e., pickup/dropoff windows, wait times) applied for trips `ContiguousZoneA` and `ZoneAtoB`, and which one consumers select depending on where the rider is going (or not going). 

That being said, some (inevitable) retooling would be required to check "one level lower" whether a rider's travel is minimally captured by a single `stop_time`.

**Comment by westontrillium** [comment]
> I hadn't considered that you would need to parse out trip_id=ZoneAtoB into separate trips for just ZoneA and just ZoneB. I would have thought that the rider could travel only within ZoneA on the ZoneAtoB trip.

Just a quick clarification (unless I'm misinterpreting your comment!), we only need to do this separation of trips because the example service we're talking about allows o/d pairs contained within each separate zone, but only o/d pairs between ZoneA and ZoneB if the pickup point is in ZoneA and drop-off point is in ZoneB. If this were a service in which you could **only** pickup in ZoneA and be dropped off in ZoneB, you _would_ just have the single...
trip_id | stop_id | stop_sequence | pickup_type | drop_off_type | start_pickup_dropoff_window | end_pickup_dropoff_window
------ | ------ | ------ | ------ | ------ | ------ | ------
ZoneAtoB | ZoneA | 1 | 2 | 1 | 08:00:00 | 17:30:00
ZoneAtoB | ZoneB | 2 | 1 | 2 | 08:00:00 | 17:30:00

and call it good.

> A huge benefit to the method you presented above disambiguates any confusion if there are different attributes (i.e., pickup/dropoff windows, wait times) applied for trips ContiguousZoneA and ZoneAtoB, and which one consumers select depending on where the rider is going (or not going).

> That being said, some (inevitable) retooling would be required to check "one level lower" whether a rider's travel is minimally captured by a single stop_time.

Always interested in learning more about the consumer side of things on spec implementation; it's important context us producers can always use more of! :)

**Comment by scmcca** [comment]
That all makes sense. Could you be more concise by doing the following:

trip_id | stop_id | stop_sequence | pickup_type | drop_off_type | start_pickup_dropoff_window | end_pickup_dropoff_window
------ | ------ | ------ | ------ | ------ | ------ | ------
ZoneAtoB | ZoneA | 1 | 2 | 2 | 08:00:00 | 17:30:00
ZoneAtoB | ZoneB | 2 | 2 | 2 | 08:00:00 | 17:30:00

This is my interpretation of this example:
- The rider can travel anywhere in ZoneA (`pickup_type=2`, `drop_off_type=2`)
- The rider can travel anywhere in ZoneB (`pickup_type=2`, `drop_off_type=2`)
- The rider can travel from ZoneA to ZoneB (from `stop_sequence=1` to `stop_sequence=2`)
- The rider cannot travel from ZoneB to ZoneA (cannot travel from `stop_sequence=2` to `stop_sequence=1` since 1 < 2, directionality is fixed)

Does that work?

**Comment by westontrillium** [comment]
That method works for your example, but not all examples unfortunately. Take a look at the following scenario:

- The rider can travel anywhere within ZoneA (`pickup_type=2`, `drop_off_type=2`) from 08:00:00 to 17:30:00
- The rider can travel anywhere within ZoneB (`pickup_type=2`, `drop_off_type=2`) **from 12:00:00 to 17:30:00**
- The rider can travel from ZoneA to ZoneB **and** from ZoneB to ZoneA from **12:00:00 to 17:30:00**

Normally, we could just put ZoneA and ZoneB in a single `location_group` to express that they're the same service area (unlimited travel within, between, etc.), but we need to parse them out because of the differing start/end_pickup_dropoff_windows. So at this point, the only things we have to tell us that travel **is** possible from ZoneB to ZoneA is that they share the same trip_id, have compatible windows, and their pickup/drop_off_types=2.

trip_id | stop_id | stop_sequence | pickup_type | drop_off_type | start_pickup_dropoff_window | end_pickup_dropoff_window
------ | ------ | ------ | ------ | ------ | ------ | ------
ZoneAandZoneB | ZoneA | 1 | 2 | 2 | 08:00:00 | 17:30:00
ZoneAandZoneB | ZoneB | 2 | 2 | 2 | 12:00:00 | 17:30:00

If we applied the `stop_sequence` directionality to this scenario, we'd be unable to express that you can pick up in ZoneB and drop off in ZoneA.

**Comment by scmcca** [comment]
I suppose this is where an additional file comes into play (i.e., wait_rules.txt in GOFS Phase 1) that can be used to nest attributes within a single stop_times.txt record to indicate non-directional service, but with different rules for different zones (such as different pickup/dropoff windows). This is out of scope for GTFS-Flex right now but could be interesting to see established. Thoughts?

**Comment by westontrillium** [comment]
So I thought about this more, and I'd like to actually recant my most recent comment! I realized I made an error in how you could represent what I thought would be a breaking example if `stop_sequence` were used in the way you proposed.

**stop_times.txt**
trip_id | stop_id | stop_sequence | pickup_type | drop_off_type | start_pickup_dropoff_window | end_pickup_dropoff_window
------ | ------ | ------ | ------ | ------ | ------ | ------
ZoneAandZoneB | ZoneA | 1 | 2 | 2 | 08:00:00 | 12:00:00
ZoneAandZoneB | Group-ZoneA-ZoneB | 2 | 2 | 2 | 12:00:00 | 17:30:00

**location_groups.txt**
location_group_id | location_id | location_group_name
----- | ----- | -----
Group-ZoneA-ZoneB | ZoneA | Springfield plus service to Shelbyville
Group-ZoneA-ZoneB | ZoneB | Springfield plus service to Shelbyville

^ Only travel within ZoneA until 12:00 when travel changes to unlimited travel within and between both zones.

So I think `stop_sequence` could actually work pretty well to express directionality! This also highlights the utility of `location_groups.txt` well.

**Comment by katibeen** [comment]
hello 
I wanted to know if the GTFS FLEX V2 file contains information about the type address to address 

Thank you
